Add `--generate-pot` option

dbnicholson opened the pull request:
POT file generation is the only part of the Gettext translation pipeline
that can't be automated. Add a --generate-pot option to run POTGenerator
without opening the editor. The editor needs to be fully initialized so
that EditorTranslationParserPlugin have been enabled.

See https://github.com/godotengine/godot-proposals/issues/10986 for background.

* Closes https://github.com/godotengine/godot-proposals/issues/10986.

dbnicholson committed:
Add --generate-pot option

POT file generation is the only part of the Gettext translation pipeline
that can't be automated. Add a --generate-pot option to run POTGenerator
without opening the editor. The editor needs to be fully initialized so
that EditorTranslationParserPlugin have been enabled.

Since the normal progress handling breaks CLI usage, change it to use text
mode for any CLI tool. Furthermore, only output the text progress in
verbose mode since they're pretty noisy.

dbnicholson commented:
@bruvzg @Calinou any thoughts on this? It would be nice to have this in 4.4.

Calinou reviewed (commented):
Tested locally (rebased on top of `master` a7a2a12bfd1c34057f158f4510841bcc766a9348), it works as expected.

Some feedback:

- Paths are relative to the project root (use an absolute path if this isn't desired). This behavior is consistent with `--export` and other arguments that accept file paths. Just a reminder for other people stumbling upon this PR :slightly_smiling_face: 
- Some verbose output is displayed when using `--generate-pot`:

```

first_scan_filesystem: begin: Project initialization steps: 5
        first_scan_filesystem: step 0: Scanning file structure...
        first_scan_filesystem: step 1: Loading global class names...
        first_scan_filesystem: step 2: Verifying GDExtensions...
        first_scan_filesystem: step 3: Creating autoload scripts...
        first_scan_filesystem: step 4: Initializing plugins...
        first_scan_filesystem: step 5: Starting file scan...
first_scan_filesystem: end
loading_editor_layout: begin: Loading editor steps: 5
        loading_editor_layout: step 0: Loading editor layout...
        loading_editor_layout: step 1: Loading docks...
        loading_editor_layout: step 2: Reopening scenes...
        loading_editor_layout: step 3: Loading central editor layout...
        loading_editor_layout: step 4: Loading plugin window layout...
        loading_editor_layout: step 5: Editor layout ready.
loading_editor_layout: end
Generating POT file out.pot
```

Can we remove it? It's not present when using `--editor` on its own. I'd expect only the `Generating POT file` line to appear.

dbnicholson commented:
> Can we remove it? It's not present when using `--editor` on its own. I'd expect only the `Generating POT file` line to appear.

Yeah, I don't like that. This is a bit of an unusual mode since the POT generator is part of the editor, so you have to load the editor to run it. I think the only other tool that does that is `--export`, and I'm pretty sure you'd see the same messages. See the first commit.

Currently it uses `print_line` for those CLI progress messages, but maybe that can be changed to `print_verbose`. I'll take a closer look.

dbnicholson commented on main/main.cpp:
```diff
@@ -3614,6 +3637,10 @@ int Main::start() {
 				patches = E->next()->get().split(",", false);
+			} else if (E->get() == "--generate-pot") {
+				ERR_FAIL_COND_V_MSG(!editor && !found_project, EXIT_FAILURE, "Please provide a valid project path when generating POT file, aborting.");
+				editor = true;
```
I think this and the above condition are wrong since `editor` is already set to true in the initial `--generator-pot` handling. I copied it from another option before tweaking the option handling later. I think it should just be `ERR_FAIL_COND_V_MSG(!found_project, EXIT_FAILURE, "Please provide a valid project path when generating POT file, aborting.");`. I'm going to take another look at this.

dbnicholson replied:
Looking again, this does appear to be correct considering the way that `found_project` and `editor` are handled in `Main::setup`. Since it matches the way the export options are handled, I'm going to leave it as I had it.

dbnicholson commented:
I added a commit to change the editor text progress messages to use `print_verbose` rather than `print_line`. This would also affect the various `--export` options, but I think it's the right thing to do as the messages fit better with the types of things you see in `--verbose` output.

I also added a fixup commit to document the `--generate-pot` option in the linux man page.

I tested locally and it does what I expect. Should I rebase the branch and/or squash the fixup?

Calinou commented:
> I tested locally and it does what I expect. Should I rebase the branch and/or squash the fixup?

Yes, please squash changes to a single commit as per the [PR workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase).

dbnicholson commented:
> Yes, please squash changes to a single commit as per the [PR workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase).

Thanks, I hadn't seen that part before. Done.

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

Note that with this PR, `--export` will no longer print a list of files when exporting unless you use `--verbose` at the same time. I personally think this is acceptable. Ideally, `--export` would print a more concise summary on its own, with only a number of files/folders (and perhaps a reminder of the non-resource export filter used, with how many files it excluded).

oxeron commented:
Hello, this looks like a great improvement, what is the next step, does it need testing ?

Calinou commented:
Commits still need to be squashed into a single commit before this can be merged. Also, merge conflicts should be resolved (the link I posted above covers thiss).

dbnicholson commented:
> Commits still need to be squashed into a single commit before this can be merged. Also, merge conflicts should be resolved ([PR workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase) covers this).

I didn't realize there was a conflict. I'll fix that up. You really want a single commit? They seem logically separate to me, but I'm happy to follow the flow.

dbnicholson commented:
> Sorry, been busy with work. I have a local build with the conflicts addressed, but I haven't had time to test it. I'll finish that up now.

Updated and tested now. It seems to work fine.

dbnicholson commented:
Rebased to fix the conflict from f11aff3841e0e721a66323bf4b3a0e71b5a14cc0.

bitterepic commented:
I hope this can get merged.  It shouldn't be a large change and it should help with automation.  I was considering implementing this myself, but I'm glad this PR exists.
